randovania/generator/hint_distributor.py

```diff
@@ -169,228 +245,224 @@ async def assign_precision_to_hints(
+
+        return patches
 
     def replace_hints_without_precision_with_jokes(self, patches: GamePatches) -> GamePatches:
         """
-        Adds WRONG_GAME precision to all hints that are missing one precision.
+        Assigns a JokeHint to each hint node with unassigned precision
         :param patches:
         :return:
         """
 
-        hints_to_replace = {asset: JokeHint() for asset, hint in patches.hints.items() if is_unassigned_location(hint)}
+        hints_to_replace = {
+            hint_node: JokeHint() for hint_node, hint in patches.hints.items() if is_unassigned_location(hint)
+        }
 
         return dataclasses.replace(
-            patches, hints={asset: hints_to_replace.get(asset, hint) for asset, hint in patches.hints.items()}
+            patches,
+            hints={hint_node: hints_to_replace.get(hint_node, hint) for hint_node, hint in patches.hints.items()},
         )
 
-    def precision_pair_weighted_list(self) -> list[PrecisionPair]:
+    @property
+    @abstractmethod
+    def default_precision_pair(self) -> PrecisionPair:
+        """The default PrecisionPair to use for unassigned generic hints."""
         raise NotImplementedError
 
-    def add_relative_hint(
+    def get_location_feature_chooser(
+        self, patches: GamePatches, location: PickupNode | None = None
+    ) -> FeatureChooser[HintFeature, HintLocationPrecision]:
+        """Create a FeatureChooser for location Features"""
+
+        region_list = patches.game.region_list
+        locations_with_feature: dict[HintFeature | HintLocationPrecision, list[PickupNode]] = defaultdict(list)
+        relevant_locations: list[PickupNode] = []
+
+        relevant_locations.extend(node for node in region_list.iterate_nodes() if isinstance(node, PickupNode))
+        for feature in patches.game.hint_feature_database.values():
+            locations_with_feature[feature].extend(region_list.pickup_nodes_with_feature(feature))
+
+        if location is not None:
+            locations_with_feature[HintLocationPrecision.REGION_ONLY] = [
+                node for node in region_list.nodes_to_region(location).all_nodes if isinstance(node, PickupNode)
+            ]
+
+        return FeatureChooser[HintFeature, HintLocationPrecision](
+            len(relevant_locations),
+            locations_with_feature,
+            HintLocationPrecision.DETAILED,
+        )
+
+    def get_pickup_feature_chooser(
         self,
-        region_list: RegionList,
-        patches: GamePatches,
+        player_pools: Sequence[PlayerPool],
+    ) -> FeatureChooser[PickupHintFeature, HintItemPrecision]:
+        """Create a FeatureChooser for pickup Features"""
+
+        pickups_with_feature: dict[PickupHintFeature | HintItemPrecision, list[PickupEntry]] = defaultdict(list)
+        relevant_pickups: list[PickupEntry] = []
+
+        for pool in player_pools:
+            for pickup in pool.pickups:
```

Comment:
Pickups already assigned to a location before the generator are ignored in this. Aka, pre-fill (DNA/etc) and pickups set to vanilla.